packages/core/echo/echo-db/src/core-db/core-database-query-context.ts

```diff
@@ -0,0 +1,104 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import { Event } from '@dxos/async';
+import { Stream } from '@dxos/codec-protobuf';
+import { Context } from '@dxos/context';
+import { log } from '@dxos/log';
+import type { QueryService, QueryResult as RemoteQueryResult } from '@dxos/protocols/proto/dxos/echo/query';
+import { nonNullable } from '@dxos/util';
+
+import type { CoreDatabase } from './core-database';
+import type { Filter } from '../query';
+import type { QueryContext, QueryResult } from '../query/query';
+
+const QUERY_SERVICE_TIMEOUT = 20_000;
+
+export class CoreDatabaseQueryContext implements QueryContext {
+  private _lastResult: QueryResult<any>[] = [];
+
+  changed = new Event();
+
+  constructor(
+    private readonly _coreDatabase: CoreDatabase,
+    private readonly _queryService: QueryService,
+  ) {}
+
+  start(): void {}
+  stop(): void {}
+
+  getResults(): QueryResult<any>[] {
+    return this._lastResult;
+  }
+
+  async run(filter: Filter<any>): Promise<QueryResult<any>[]> {
+    const queryId = nextQueryId++;
+    // Disposed when this method exists.
+    await using ctx = new Context();
+
+    const start = Date.now();
+
+    const response = await Stream.first(
+      this._queryService.execQuery({ filter: filter.toProto() }, { timeout: QUERY_SERVICE_TIMEOUT }),
+    );
+
+    if (!response) {
+      throw new Error('Query terminated without a response');
```

Comment:
I think we're capitalizing errors with a period where appropriate (unlike logging which is all lower case, no punctuation).